src/sinks/util/tls.rs

```diff
@@ -0,0 +1,192 @@
+    x509::X509,
+};
+use serde::{Deserialize, Serialize};
+use snafu::{ResultExt, Snafu};
+use std::convert::TryFrom;
+use std::fmt::Debug;
+use std::fs::File;
+use std::io::Read;
+use std::path::{Path, PathBuf};
+
+#[derive(Debug, Snafu)]
+enum TlsError {
+    #[snafu(display("Could not open {} file {:?}: {}", note, filename, source))]
+    FileOpenFailed {
+        note: &'static str,
+        filename: PathBuf,
+        source: std::io::Error,
+    },
+    #[snafu(display("Could not read {} file {:?}: {}", note, filename, source))]
+    FileReadFailed {
+        note: &'static str,
+        filename: PathBuf,
+        source: std::io::Error,
+    },
+    #[snafu(display("Could not set TCP TLS identity: {}", source))]
+    TlsIdentityError { source: native_tls::Error },
+    #[snafu(display("Could not export identity to DER: {}", source))]
+    DerExportError { source: openssl::error::ErrorStack },
+    #[snafu(display("Could not parse certificate in {:?}: {}", filename, source))]
+    CertificateParseError {
+        filename: PathBuf,
+        source: native_tls::Error,
+    },
+    #[snafu(display("Must specify both TLS key_file and crt_file"))]
+    MissingCrtKeyFile,
+    #[snafu(display("Could not parse X509 certificate in {:?}: {}", filename, source))]
+    X509ParseError {
+        filename: PathBuf,
+        source: openssl::error::ErrorStack,
+    },
+    #[snafu(display("Could not parse private key in {:?}: {}", filename, source))]
+    PrivateKeyParseError {
+        filename: PathBuf,
+        source: openssl::error::ErrorStack,
+    },
+    #[snafu(display("Could not build PKCS#12 archive for identity: {}", source))]
+    Pkcs12Error { source: openssl::error::ErrorStack },
+}
+
+/// Standard TLS connector options
+#[derive(Debug, Clone, Derivative, Deserialize, Serialize)]
+#[derivative(Default)]
+pub struct TlsOptions {
+    #[derivative(Default(value = "true"))]
+    pub verify_certificate: bool,
+    pub ca_path: Option<PathBuf>,
+    pub crt_path: Option<PathBuf>,
+    pub key_path: Option<PathBuf>,
+    pub key_pass: Option<String>,
+}
+
+/// Directly usable settings for TLS connectors
+#[derive(Derivative)]
+#[derivative(Default)]
+pub struct TlsSettings {
+    #[derivative(Default(value = "true"))]
+    verify_certificate: bool,
+    authority: Option<Certificate>,
+    identity: Option<Identity>,
+}
+
+impl TryFrom<&TlsOptions> for TlsSettings {
+    type Error = crate::Error;
+    fn try_from(options: &TlsOptions) -> Result<Self, Self::Error> {
+        if options.crt_path.is_some() != options.key_path.is_some() {
+            return Err(TlsError::MissingCrtKeyFile.into());
+        }
+
+        let authority = match options.ca_path {
+            None => None,
+            Some(ref path) => Some(load_certificate(path)?),
+        };
+
+        let identity = match options.crt_path {
+            None => None,
+            Some(ref crt_path) => {
+                let identity = load_build_pkcs12(
```

Comment:
since we are making this generic we prob want to start thinking about other formats like pkcs8 and rsa?

Reply:
Absolutely. Natively supporting PKCS#12 identity files instead of creating them from a pair of PEM files would be a first step. Probably should go in a separate PR though.

Reply:
Yup, just thinking out loud a bit. We should make a follow-up issue for this.